letta/services/passage_manager.py

```diff
@@ -0,0 +1,225 @@
+        """Create a new passage."""
+        with self.session_maker() as session:
+            passage = PassageModel(**pydantic_passage.model_dump())
+            passage.create(session, actor=actor)
+            return passage.to_pydantic()
+
+    @enforce_types
+    def create_many_passages(self, passages: List[PydanticPassage], actor: PydanticUser) -> List[PydanticPassage]:
+        """Create multiple passages."""
+        return [self.create_passage(p, actor) for p in passages]
+
+    @enforce_types
+    def insert_passage(self, 
+        agent_state: AgentState,
+        agent_id: str,
+        text: str, 
+        actor: PydanticUser, 
+        return_ids: bool = False
+    ) -> List[PydanticPassage]:
+        """ Insert passage(s) into archival memory """
+
+        embedding_chunk_size = agent_state.embedding_config.embedding_chunk_size
+        embed_model = embedding_model(agent_state.embedding_config)
+
+        passages = []
+        
+        try:
+            # breakup string into passages
+            for text in parse_and_chunk_text(text, embedding_chunk_size):
+                embedding = embed_model.get_text_embedding(text)
+                if isinstance(embedding, dict):
+                    try:
+                        embedding = embedding["data"][0]["embedding"]
+                    except (KeyError, IndexError):
+                        # TODO as a fallback, see if we can find any lists in the payload
+                        raise TypeError(
+                            f"Got back an unexpected payload from text embedding function, type={type(embedding)}, value={embedding}"
+                        )
+                passage = self.create_passage(
+                    PydanticPassage(
+                        organization_id=actor.organization_id,
+                        agent_id=agent_id,
+                        text=text,
+                        embedding=embedding,
+                        embedding_config=agent_state.embedding_config
+                    ),
+                    actor=actor
+                )
+                passages.append(passage)
+
+            ids = [str(p.id) for p in passages]
+
+            if return_ids:
+                return ids
+            
+            return passages
+
+        except Exception as e:
+            raise e
+
+    @enforce_types
+    def update_passage_by_id(self, passage_id: str, passage: PydanticPassage, actor: PydanticUser, **kwargs) -> Optional[PydanticPassage]:
+        """Update a passage."""
+        if not passage_id:
+            raise ValueError("Passage ID must be provided.")
+
+        with self.session_maker() as session:
+            try:
+                # Fetch existing message from database
+                curr_passage = PassageModel.read(
+                    db_session=session,
+                    identifier=passage_id,
+                    actor=actor,
+                )
+                if not curr_passage:
```

Comment:
read will error here, it will not return None - I think you are thinking of your manager function

Reply:
discussed offline, will do in a follow up PR